crates/mod-rspamd/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+#[derive(Debug, Default, Clone, mlua::FromLua)]
+struct EnvelopeDataLua {
+    from: Option<String>,
+    rcpt: Vec<String>,
+    ip: Option<String>,
+    user: Option<String>,
+    helo: Option<String>,
+    hostname: Option<String>,
+    queue_id: Option<String>,
+    file_path: Option<String>,
+    body_block: bool,
+    additional_headers: Option<HashMap<String, String>>,
+}
+
+impl EnvelopeDataLua {
+    fn into_envelope_data(self) -> EnvelopeData {
+        // Build additional headers, adding queue_id if present
+        let mut additional_headers = self.additional_headers.unwrap_or_default();
+        if let Some(queue_id) = self.queue_id {
+            additional_headers.insert("Queue-Id".to_string(), queue_id);
+        }
+
+        // Construct EnvelopeData directly since all fields are public
+        EnvelopeData {
+            from: self.from,
+            rcpt: self.rcpt,
+            ip: self.ip,
+            user: self.user,
+            helo: self.helo,
+            hostname: self.hostname,
+            file_path: self.file_path,
+            body_block: self.body_block,
+            additional_headers,
+        }
+    }
+}
+
+/// Register the rspamd module with Lua
+pub fn register(lua: &Lua) -> anyhow::Result<()> {
+    let rspamd_mod = get_or_create_sub_module(lua, "rspamd")?;
+
+    rspamd_mod.set(
+        "build_client",
+        lua.create_async_function(|lua, options: Value| async move {
+            let config: RspamdClientConfig = from_lua_value(&lua, options)?;
+            RspamdClient::new(config).await.map_err(any_err)
+        })?,
+    )?;
+
+    rspamd_mod.set(
+        "scan_message",
+        lua.create_async_function(|lua, (config_value, msg): (Value, Message)| async move {
+            let config: RspamdClientConfig = from_lua_value(&lua, config_value)?;
+
+            // Extract envelope data from message
+            let envelope_data = EnvelopeData {
+                from: msg.sender().map(|s| s.to_string()).ok(),
+                rcpt: msg.recipient_list_string().unwrap_or_default(),
+                ip: msg.get_meta_string("received_from").ok().flatten(),
+                user: msg.get_meta_string("authn_id").ok().flatten(),
+                helo: msg.get_meta_string("ehlo_domain").ok().flatten(),
+                hostname: msg.get_meta_string("hostname").ok().flatten(),
+                // We don't have a reliable file path in most cases
+                file_path: None,
+                body_block: false,
+                additional_headers: HashMap::default(),
+            };
+
+            // Build client config and scan
+            // scan_async accepts Into<Bytes>, so we convert Arc<Box<[u8]>> to Vec<u8>
+            // which then gets wrapped in Bytes. This minimizes intermediate allocations.
+            let client_config = config.make_client_config().await.map_err(any_err)?;
+            let reply = scan_async(&client_config, (*msg.get_data()).to_vec(), envelope_data)
```

Comment:
Should be able to do something like this to get zero-copy of the data bytes:

```suggestion
           struct DataWrapper(Arc<Box<[u8]>>);
           impl AsRef<[u8]> for DataWrapper {
               fn as_ref(&self) -> &[u8] {
                   &*self.0
                }
            } 
            let data = Bytes::from_owner(DataWrapper(msg.get_data()));

            let client_config = config.make_client_config().await.map_err(any_err)?;
            let reply = scan_async(&client_config, data, envelope_data)
```

Reply:
I need to update rspamdclient-rs to make it happen as well.